### PART - I

Q1.

a. Given the string 'great', use an appropriate index command that returns 'a'
from the string. Given s = ‘great’. (0.5 point)
Expected output: ‘a’

In [1]:
s='great'
print(s[3])

a


b. Write a program to find out the second smallest number in the list given
below. (0.5 point)
li=[32, 3, 5, 8, 33, 7] & Expected output: 5


In [2]:
li = [32,3,5,8,33,7]
li.sort()
print(li[1])

5


c. Find out the unique values of the list given below. (0.5 point)
mylist=[1,1,5,3,3,6,11,11] & Expected output: [1,5,3,6,11]

In [3]:
mylist = [1,1,5,3,3,6,11,11]

ans = []

for element in mylist:
    
    if element not in ans:
        ans.append(element)
        
print(ans)

[1, 5, 3, 6, 11]


d. Write a program to print cube root of 216 using appropriate syntax. (0.5
point)
Expected output: 6

In [4]:
x=216
ans= round(pow(x,1/3))
print(ans)

6


Q2. Replace the value 'hello' in the below nested list with 'bye'. (1 point)

a = [1, 2, [3, 4, 'hello']] & Expected output: [1, 2, [3, 4, 'bye']]

In [5]:
a = [1, 2, [3, 4, 'hello']]
a[2][2]='bye'
print(a)

[1, 2, [3, 4, 'bye']]


Q3. Given the below dictionary, write a program that returns [‘object0’]. (1 point)
dict = {'a':9, 'b':(2,3,1), 'c':['object0','object1','object2']}
Expected output: [‘object0’]

In [6]:
dict = {'a':9, 'b':(2,3,1), 'c':['object0','object1','object2']}
dict['c'][0]

'object0'

Q4. Write a program to calculate the area of a circle. The program should ask for an
appropriate input and print the computed area. (2 points)

a. Input text: Please enter the radius of the circle.

b. If the user enters a value of 5. The output should be: The area of the circle
is 78.54

c. This program should indicate that the input is invalid if the user enters a
character instead of a number as input. For instance if the user enters a
value of ‘a’ instead of 5 in the previous example the program should
prompt: Invalid input, please enter a number: _

In [7]:
import math

try:
    radius=float(input("Please enter the radius of the circle"))
    area = math.pi*radius*radius
    print(f"The area of the circle is {round(area,2)}")
except ValueError:
    print("Invalid input, please enter a number: _")   

Please enter the radius of the circle5
The area of the circle is 78.54


Q5. Write a program which will take two lists X & Y as input and it will return a list
having all the odd numbers present in the two lists. (Note:- common elements
can be repeated) (2 points)

Input list: X=[1,2,3,4,5] & Y=[5,6,7,8,9]

Expected Output: [1,3,5,5,7,9] or [1,3,5,7,9] both are accepted.

In [28]:
x = [int(x) for x in input("X=").split(",")]
y = [int(y) for y in input("Y=").split(",")]
output_li=[]
def oddlist (x, y):
    x.extend(y)
    for val in x:
        if (val%2!=0):
            output_li.append(val)
    return output_li

oddlist(x,y)

X=1,2,3,4,5
Y=5,6,7,8,9


[1, 3, 5, 5, 7, 9]

Q6. Write a function that prints the integers from 1 to 15. But for multiples of three
print "GREAT" instead of the number, and for the multiples of five print
"LEARNING". For numbers which are multiples of both three and five print
"GREATLEARNING". (2 points)

In [40]:
def printpattern():
    
    for num in range(1,16):
        if(num%3==0 and num%5==0):
            print("GREATLEARNING")
        elif(num%3==0):
            print("GREAT")
        elif(num%5==0):
            print("LEARNING")
        else:
            print(num)
        
printpattern()

1
2
GREAT
4
LEARNING
GREAT
7
8
GREAT
LEARNING
11
GREAT
13
14
GREATLEARNING


### PART - II

Dataset: Online_sales.csv

Domain: Retail

Objective: Analyzing sales data to understand sales trend

1. Import necessary libraries and read the provided dataset (online_sales.csv) and
check the top 5 and random 5 samples of the dataframe. (0.5 point)

2. Check info of the dataframe and write your observations. Comment on data
types and shape of the dataset. (0.5 point)

3. Check for null values and report the percentage of null values of each column.
And drop the rows having null values in it. (1 point)

4. Check the statistical summary of the dataset. (Use inbuilt functions) (0.5 point)

5. Drop the instances having quantity less than zero. (1 point) (Note:- refer to the
‘Quantity’ feature)

6. Check unique values of the country and report the name of the country that has
the highest number of instances/rows. (0.5 point)

7. Create a new column with the name as ‘sales’ having total sales. The total sales
is defined as Quantity*UnitPrice. (1 point)

8. Report the top 5 countries in terms of sales. (2 points)

   a. Consider the size of sales.
   
   b. Consider the mean value of sales.
   

9. Report the top 5 products which bring the highest sales. Use StockCode for
product information. (1 point)

10. Convert the ‘InvoiceDate’ into a date format and report the month on which
the maximum sales occur? (2 points)

Solutions to the above questions:

In [41]:
import numpy as np
import pandas as pd

In [44]:
df=pd.read_csv("/Users/simrannarora/Downloads/online_sales.csv")

In [45]:
df.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

    InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/10 8:26       2.55     17850.0  United Kingdom  
1  12/1/10 8:26       3.39     17850.0  United Kingdom  
2  12/1/10 8:26       2.75     17850.0  United Kingdom  
3  12/1/10 8:26       3.39     17850.0  United Kingdom  
4  12/1/10 8:26       3.39     17850.0  United Kingdom

In [50]:
df.sample(n=5)

InvoiceNo StockCode                      Description  Quantity  \
57753     541219     22834      HAND WARMER BABUSHKA DESIGN         6   
174028    551734     22951  60 CAKE CASES DOLLY GIRL DESIGN        24   
105508    545227     22052        VINTAGE CARAVAN GIFT WRAP        25   
95398     544455     22625               RED KITCHEN SCALES         1   
88381     543802     21754         HOME BUILDING BLOCK WORD         8   

          InvoiceDate  UnitPrice  CustomerID         Country  
57753   1/14/11 14:06       1.63         NaN  United Kingdom  
174028   5/4/11 10:15       0.55     18178.0  United Kingdom  
105508    3/1/11 9:35       0.42     12428.0         Finland  
95398   2/20/11 13:18       8.50     13097.0  United Kingdom  
88381   2/13/11 12:03       5.95     13742.0  United Kingdom

In [51]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 240007 entries, 0 to 240006
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    240007 non-null  object 
 1   StockCode    240007 non-null  object 
 2   Description  239106 non-null  object 
 3   Quantity     240007 non-null  int64  
 4   InvoiceDate  240007 non-null  object 
 5   UnitPrice    240007 non-null  float64
 6   CustomerID   172782 non-null  float64
 7   Country      240007 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 14.6+ MB


Observations - 

- The class of the above dataframe is 'pandas.core.frame.DataFrame'.
- There are total 240007 entries with indexes ranging from 0 to 240006.
- There are total 8 columns in this dataset.
- The number of columns along with column name is listed in the above table. The Non-Null Count here shows the number   of non-null values in each column along with the datatype of each column.
- Dtypes show all the data types that are there in the dataset and the number of columns having that particular       datatype ( float64(2), int64(1), object(5)).
- Memory usage shows the memory space that is taken by the dataset ( 14.6 MB)

In [53]:
df.dtypes

InvoiceNo       object
StockCode       object
Description     object
Quantity         int64
InvoiceDate     object
UnitPrice      float64
CustomerID     float64
Country         object
dtype: object

The datatypes of this dataset are:
- float64 - 2 columns
- int64 - 1 column
- object - 5 columns

In [54]:
df.shape

(240007, 8)

The shape of the dataframe is represented in tuple format where 240007 are the no. of rows and 8 are the no. of columns in the dataset

In [56]:
df.isnull().sum()

InvoiceNo          0
StockCode          0
Description      901
Quantity           0
InvoiceDate        0
UnitPrice          0
CustomerID     67225
Country            0
dtype: int64

In [57]:
print(df.isnull().values.sum()) # This is the no. of missing values

68126


In [60]:
df.isnull().mean()*100

InvoiceNo       0.000000
StockCode       0.000000
Description     0.375406
Quantity        0.000000
InvoiceDate     0.000000
UnitPrice       0.000000
CustomerID     28.009600
Country         0.000000
dtype: float64

In [64]:
df=df.dropna()

In [65]:
df.isnull().sum()

InvoiceNo      0
StockCode      0
Description    0
Quantity       0
InvoiceDate    0
UnitPrice      0
CustomerID     0
Country        0
dtype: int64

In [66]:
df.describe()

Quantity      UnitPrice     CustomerID
count  172782.000000  172782.000000  172782.000000
mean       12.244835       3.793566   15274.819941
std       257.898235     101.069930    1725.093177
min    -74215.000000       0.000000   12346.000000
25%         2.000000       1.250000   13842.000000
50%         5.000000       1.950000   15132.000000
75%        12.000000       3.750000   16814.000000
max     74215.000000   38970.000000   18287.000000

In [79]:
dropval=df[df['Quantity'] < 0].index
df.drop(dropval,inplace=True)

In [80]:
df

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
240002    558077    47590A          BLUE HAPPY BIRTHDAY BUNTING        12   
240003    558077     82600                NO SINGING METAL SIGN        12   
240004    558077    85123A   WHITE HANGING HEART T-LIGHT HOLDER        32   
240005    558077     22746        POPPY'S PLAYHOUSE LIVINGROOM         12   
240006    558077     23251         VINTAGE RED ENAMEL TRIM MUG         12   

          InvoiceDate  UnitPrice  CustomerID         Country  
0        12/1/10 8:26       2.55     17850.0  United Kingdom  
1        12/1/10 8:26       3.39     17850.0  United Kingdom  
2        12/1/10 8:26       2.75     17850.0  United Kingdom  
3        12/1/10 8:26       3.39     17850.0  United Kingdom  
4        12/1/10 8:26       3.39     17850.0  United Kingdom  
...               ...        ...         ...             ...  
240002  6/26/11 10:59       5.45     13408.0  United Kingdom  
240003  6/26/11 10:59       2.10     13408.0  United Kingdom  
240004  6/26/11 10:59       2.55     13408.0  United Kingdom  
240005  6/26/11 10:59       2.10     13408.0  United Kingdom  
240006  6/26/11 10:59       1.25     13408.0            Unit  

[168675 rows x 8 columns]

In [84]:
df['Country'].unique()

array(['United Kingdom', 'France', 'Australia', 'Netherlands', 'Germany',
       'Norway', 'EIRE', 'Switzerland', 'Spain', 'Poland', 'Portugal',
       'Italy', 'Belgium', 'Lithuania', 'Japan', 'Iceland',
       'Channel Islands', 'Denmark', 'Cyprus', 'Sweden', 'Finland',
       'Austria', 'Greece', 'Singapore', 'Lebanon',
       'United Arab Emirates', 'Israel', 'Saudi Arabia', 'Czech Republic',
       'Canada', 'Unspecified', 'Brazil', 'USA', 'European Community',
       'Bahrain', 'Malta', 'Unit'], dtype=object)

In [82]:
df['Country'].nunique()

37

In [85]:
item_counts = df["Country"].value_counts()
print(item_counts)

United Kingdom          150046
Germany                   3958
France                    3562
EIRE                      2650
Netherlands               1139
Spain                     1131
Belgium                    914
Switzerland                697
Australia                  630
Portugal                   610
Norway                     368
Channel Islands            363
Cyprus                     350
Finland                    307
Italy                      299
Japan                      230
Sweden                     195
Denmark                    184
Poland                     180
Austria                    124
Singapore                  113
Iceland                    102
Greece                      85
Unspecified                 72
Canada                      68
Lebanon                     45
Malta                       45
Lithuania                   35
Brazil                      32
European Community          31
United Arab Emirates        30
USA                         22
Bahrain 

This shows that United Kingdom has highest number of instances/rows

In [88]:
df["Sales"] = df["Quantity"] * df["UnitPrice"]

In [89]:
df.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

    InvoiceDate  UnitPrice  CustomerID         Country  Sales  
0  12/1/10 8:26       2.55     17850.0  United Kingdom  15.30  
1  12/1/10 8:26       3.39     17850.0  United Kingdom  20.34  
2  12/1/10 8:26       2.75     17850.0  United Kingdom  22.00  
3  12/1/10 8:26       3.39     17850.0  United Kingdom  20.34  
4  12/1/10 8:26       3.39     17850.0  United Kingdom  20.34

In [90]:
df["Sales"].mean()

23.05600557877644

In [92]:
df["Sales"].count()

168675

In [103]:
df.groupby('Country')['Sales'].sum().nlargest(5) #considering size of sales

Country
United Kingdom    3167696.671
Netherlands        125816.110
Germany            106223.190
EIRE               101406.060
France              89349.880
Name: Sales, dtype: float64

In [104]:
df.groupby('Country')['Sales'].mean().nlargest(5) #considering mean of the sales

Country
Australia      126.677444
Netherlands    110.461905
Japan          100.181609
Singapore       90.819912
Sweden          86.532205
Name: Sales, dtype: float64

In [111]:
df.groupby('StockCode')['Sales'].sum().nlargest(5)

StockCode
22423     84734.20
23166     78499.90
85123A    58904.07
22502     47167.85
47566     40513.85
Name: Sales, dtype: float64

In [112]:
df['InvoiceDate'] = pd.to_datetime(df['InvoiceDate'])

In [113]:
df.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  Sales  
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom  15.30  
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  20.34  
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom  22.00  
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  20.34  
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  20.34

In [124]:
monthly_sales = df.groupby(df['InvoiceDate'].dt.month)['Sales'].sum()

In [125]:
max_month = monthly_sales.idxmax()

In [126]:
max_month

5

May is the month where maximum sales occur